include/seqan3/index/fm_index_iterator.hpp

```diff
@@ -0,0 +1,536 @@
+ *
+ * The asymptotic running times for using the iterator depend on the SDSL index configuration. To determine the exact
+ * running times, you have to additionally look up the running times of the used traits (configuration).
+ */
+template <typename index_t>
+class fm_index_iterator
+{
+
+public:
+    //!\brief Type of the index.
+    using index_type = index_t;
+    //!\brief Type for representing positions in the indexed text.
+    using size_type = typename index_type::size_type;
+    // using text_pos_t = typename index_t::size_type; // depends on dimensions of the text
+
+protected:
+    //!\privatesection
+
+    //!\brief Type of the representation of a suffix tree node.
+    using node_type = detail::fm_index_iterator_node<index_t>;
+
+    //!\brief Type of the underlying FM index.
+    const index_type * index;
+    //!\brief Left suffix array range of the parent node. Needed for right().
+    size_type parent_lb;
+    //!\brief Right suffix array range of the parent node. Needed for right().
+    size_type parent_rb;
+    //!\brief Underlying index from the SDSL.
+    node_type node;
+
+    //!\brief Helper function to recompute text positions since the indexed text is reversed.
+    size_type offset() const noexcept
+    {
+        return index->m_index.size() - depth() - 1; // since the string is reversed during construction
+    }
+
+    //!\publicsection
+
+public:
+
+    // TODO(h-2):
+    // * shouldn't all these constructors be noexcept? construction will happen a lot in recursions
+    // * better (more generic) explanation why the default constructor is necessary
+    //!\brief Default constructor. Accessing member functions on a default constructed object is undefined behavior.
+    //        Default construction is necessary for example to construction std::array of iterators.
+    fm_index_iterator() = default;
+    /*!\name Constructors and destructor
+     * \{
+     */
+    fm_index_iterator(fm_index_iterator const &) = default;
+    fm_index_iterator & operator=(fm_index_iterator const &) = default;
+    fm_index_iterator(fm_index_iterator &&) = default;
+    fm_index_iterator & operator=(fm_index_iterator &&) = default;
+
+    fm_index_iterator(index_t const & _index) : index(&_index), node({0, _index.m_index.size() - 1, 0, 0})
+    {}
+    //\}
+
+    /*!\brief Compares two iterators.
+     * \param[in] rhs Other iterator to compare it to.
+     * \returns `true` if both iterators are equal, `false` otherwise.
+     *
+     * ### Complexity
+     *
+     * Constant. (TODO: will change when comparing indices, not yet supported by SDSL)
+     *
+     * ### Exceptions
+     *
+     * No-throw guarantee.
+     */
+    bool operator==(fm_index_iterator const & rhs) const noexcept
+    {
+        assert(index != nullptr);
+
+        // parent_lb/parent_rb might be uninitialized in a root node
+        return node == rhs.node && ((parent_lb == rhs.parent_lb && parent_rb == rhs.parent_rb) || depth() == 0);
```

Comment:
check `depth == 0` before the other condition to prevent read from unitialised memory